### ISSUE TITLE
Unbound pythonmod crashes with HUP

### ISSUE THREAD

**Issue by BBcan177** [NLnetLabs/unbound#372]
With the pythonmod feature enabled, and a python script running, sending a **HUP** command or attempting a "**reload**" with **unbound-control** will cause Unbound to crash.

```
in method '_get_qname', argument 1 of type 'struct query_info *'
in method 'module_qstate_qinfo_get', argument 1 of type 'struct module_qstate *'
```

**Comment by rcmcdonald91** [comment]
I'm seeing this with when using unbound 1.17 and python3.9 both built with debug symbols.

https://github.com/python/cpython/blob/e8f61ede93ecbda3b2055a2e1d8c0b43b2049737/Objects/typeobject.c#L3487

```
Objects/typeobject.c:3487: type_dealloc: Assertion "type->tp_flags & (1UL << 9)" failed
Enable tracemalloc to get the memory block allocation traceback

object address  : 0x844928b50
object refcount : 0
object type     : 0x825897f08
object type name: type
object repr     : <refcnt 0 at 0x844928b50>

Fatal Python error: _PyObject_AssertFailed: _PyObject_AssertFailed
Python runtime state: finalizing (tstate=0x82b8badd0)

Current thread 0x000000082b754000 (most recent call first):
<no Python frame>
Abort
```

It seems like some heap-allocated object is being free'd by the garbage collector when it should not be. Do we need to bump a refcount somewhere?

**Comment by rcmcdonald91** [comment]
We can close this.

The issue that @BBcan177 and I were observing was traced back to the MaxMindDB Python Module (yes, it was indeed missing a `Py_INCREF` as I suspected).

In debugging this problem though I found several points of improvement to the Python module, most notably we should not be reloading the interpreter each time we reload unbound (either via SIGHUP or `unbound-control reload`). Reloading the interpreter via `Py_Initialize/Py_Finalize` cycles is problematic, especially on Python 3.9. Memory management has vastly improved in Python 3.10+. This is especially useful for consumers embedding the python interpreter (as-is Unbound).

As an example, start up unbound compiled against python 3.9 and HUP 'bomb' the process... the maxrss explodes pretty quickly. Obviously this is a worst case, but there are indeed consumers of Unbound that frequently SIGHUP or reload the running Unbound process in order to modify the running state.
